tatsumi - begin some refactoring

mamehaze opened the pull request:
- split rotating sprites into a device
- split into 3 drivers (apache3, roundup5, cyclwarr) as apart from sprites they have little actually in common
- a few small cleanups

mamehaze commented:
note, a lot of the *code* here still isn't very good, this is mostly a restructuring to make future work easier / more manageable, it isn't a code clean-up pass as such.

I do plan on looking at things like the sprite rotation again at some point, I did have some old progress but appear to have lost it, as I tried to take on too much at once without committing anything, rather than submit a smaller part as I'm doing here.

cuavas commented on src/mame/tatsumi/cyclwarr.cpp:
```diff
@@ -0,0 +1,1286 @@
+#include "speaker.h"
+
+/* TODO:
+    - (fixed) Cycle Warriors: transparent road layer on sidelines, wrong mask_data?
+    - (fixed) Missing BG layer (Round Up 5) - banked VRAM data from somewhere!?
+    - (fixed?) Cycle Warriors: test mode text does not appear as it needs a -256 Y
+      scroll offset from somewhere.
+    - (fixed) Cycle Warriors: sometimes it draws garbage on character select or even hangs
+      depending on where player coins up, most likely caused by miscommunication with sub CPU?
+    - (fixed) Cycle Warriors: ranking screen is completely wrong;
+    - (fixed) Cycle Warriors: ugly orange color on character select and briefing screens, layer disable?
```
There’s no point keeping the stuff that’s done in the TODO list…

modusc896d352 replied:
do keep the (fixed?) comment though, you never know when that'll turn out to be a massive oversight.

angelosa replied:
iirc text layer should come from 6845 so it's something that needs doing anyway, the note is safe to remove imo.

angelosa replied:
Actually nm: this is `tatsumi/cyclwarr.cpp` not `roundup5.cpp`, don't remember what's that for then.

mamehaze replied:
once again, this is just from the existing files, the goal here was to preserve what was there, if a decision is made to not want that information it's really something entirely separate from this PR which you just seem to be piling additional requests onto outside the scope of what's being done.

cuavas commented on src/mame/tatsumi/cyclwarr.cpp:
```diff
@@ -0,0 +1,1286 @@
+	// bit 14: ignore transparency on this tile
+	int opaque = ((tile >> 14) & 1) == 1;
+
+	tileinfo.set(0,
+						 tileno,
+						 color,
+						opaque ? TILE_FORCE_LAYER0 : 0);
```
Indentation is all over the place here.

mamehaze replied:
again this hasn't actually changed from MAME, please keep the scope of the feedback to things I've actually changed.

cuavas commented on src/mame/tatsumi/cyclwarr.cpp:
```diff
@@ -0,0 +1,1286 @@
+	color |= 4;
+	int opaque = ((tile >> 14) & 1) == 1;
+
+	tileinfo.set(0,
+						 tileno,
+						 color | m_road_color_bank,
+						 opaque ? TILE_FORCE_LAYER0 : 0);
```
Same here.

cuavas commented on src/mame/tatsumi/roundup5.cpp:
```diff
@@ -0,0 +1,783 @@
+	void gfxdata_w(offs_t offset, uint8_t data);
+	void output_w(uint8_t data);
+
+	DECLARE_VIDEO_START(roundup5);
+	uint32_t screen_update_roundup5(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect);
```
You should just override `video_start()` for this derived class rather than using the legacy `VIDEO_START` stuff.

mamehaze replied:
This hasn't changed from what's already in MAME, again not really eligible for PR review here.  I may change it anyway, but again please concentrate on what I have changed, picking up on things I haven't changed at all is not helpful.

cuavas commented on src/mame/tatsumi/roundup5.cpp:
```diff
@@ -0,0 +1,783 @@
+
+	    0x0017  :   OBJ banks
+	    0x0018  :   68000 RAM       mask 0x0380 used to save bits when writing
+	    0x0c10  :   68000 ROM
+
+	    0x0040  :   Z80 rom (lower half) mapped to 0x10000
+	    0x0060  :   Z80 rom (upper half) mapped to 0x10000
+
+	    0x0080  :   enabled when showing map screen after a play
+	                (switches video priority between text layer and sprites)
+
+	    0x0100  :   watchdog.
+
+	    0x0c00  :   vram bank (bits 0x7000 also set when writing vram)
+
+	    0x8000  :   set whenever writing to palette ram?
+
+	    Changed bank to 0060 (0)
+	*/
+
+	if ((m_control_word & 0x8) == 0 && !(m_last_control & 0x8))
+		m_subcpu->set_input_line(INPUT_LINE_IRQ4, ASSERT_LINE);
+//  if (tatsumi_control_w&0x200)
+//      cpu_set_reset_line(1, CLEAR_LINE);
+//  else
+//      cpu_set_reset_line(1, ASSERT_LINE);
+
+//  if ((tatsumi_control_w&0x200) && (last_bank&0x200)==0)
+//      logerror("68k irq\n");
+//  if ((tatsumi_control_w&0x200)==0 && (last_bank&0x200)==0x200)
+//      logerror("68k reset\n");
+
+	if (m_control_word == 0x3a00)
+	{
+//      cpu_set_reset_line(1, CLEAR_LINE);
+//      logerror("68k on\n");
+	}
+
+	m_last_control = m_control_word;
+}
```
This is pretty gross.

`m_last_control` will always equal `m_control_word` outside this function, so it’s pointless making both of them members.  It would be more sensible to make the new control word value a local variable, and update `m_control_word` at the end.

Also, it’s going to be going through the scheduler every time the halt lines for `m_audiocpu` and `m_subcpu` *don’t* change because it isn’t checking that they’ve actually changed.

mamehaze replied:
this is existing code, it is no better or worse than what's in MAME right now because it is what's in MAME right now, this is not eligible for PR review.  Changes to things like this will be dealt with at a later date, in a future PR once I intend to make functional changes.

cuavas commented on src/mame/tatsumi/tatsumi_rotating_sprites.h:
```diff
@@ -0,0 +1,84 @@
+	void draw_sprites(bitmap_ind8& bitmap, const rectangle& cliprect, int write_priority_only, int rambank);
+
+	void update_cluts();
+
+	DECLARE_GFXDECODE_MEMBER(gfxinfo);
```
Does this need to be public?

cuavas commented on src/mame/tatsumi/tatsumi_rotating_sprites.h:
```diff
@@ -0,0 +1,84 @@
+	required_device<palette_device> m_basepalette;
+
+private:
+	std::unique_ptr<uint8_t[]> m_shadow_pen_array;
+
+	bitmap_rgb32 m_temp_bitmap;
+
+	int m_sprite_palette_base;
+
+	void mycopyrozbitmap_core(bitmap_ind8 &bitmap, const bitmap_rgb32 &srcbitmap,
+			int dstx, int dsty, int srcwidth, int srcheight, int incxx, int incxy, int incyx, int incyy,
+			const rectangle &clip, int transparent_color);
+	void mycopyrozbitmap_core(bitmap_rgb32 &bitmap, const bitmap_rgb32 &srcbitmap,
+			int dstx, int dsty, int srcwidth, int srcheight, int incxx, int incxy, int incyx, int incyy,
+			const rectangle &clip, int transparent_color);
+
+	template<class BitmapClass> void draw_sprites_main(BitmapClass &bitmap, const rectangle &cliprect, int write_priority_only, int rambank);
+	template<class BitmapClass> inline void roundupt_drawgfxzoomrotate( BitmapClass &dest_bmp, const rectangle &clip,
+		gfx_element *gfx, uint32_t code,uint32_t color,int flipx,int flipy,uint32_t ssx,uint32_t ssy,
+		int scalex, int scaley, int rotate, int write_priority_only );
+
+	required_shared_ptr<uint16_t> m_spriteram;
+	required_region_ptr<uint8_t> m_sprites_l_rom;
+	required_region_ptr<uint8_t> m_sprites_h_rom;
```
Please keep the data members together and the member functions together.

cuavas commented on src/mame/tatsumi/tatsumi_rotating_sprites.cpp:
```diff
@@ -0,0 +1,552 @@
+{
+}
+
+void tatsumi_rotating_sprites_device::device_add_mconfig(machine_config &config)
+{
+	PALETTE(config, m_fakepalette).set_format(palette_device::xRGB_555, 4096); // 4096 arranged as series of CLUTs
+}
+
+void tatsumi_rotating_sprites_bigpal_device::device_add_mconfig(machine_config &config)
+{
+	PALETTE(config, m_fakepalette).set_format(palette_device::xRGB_555, 8192); // 8192 arranged as series of CLUTs
+}
```
You don’t need to override this in the derived class, you can just use `m_rom_clut_size * 2` for the number of entries.

cuavas commented on src/mame/tatsumi/tatsumi_rotating_sprites.h:
```diff
@@ -0,0 +1,84 @@
+	virtual void device_start() override ATTR_COLD;
+	virtual void device_reset() override ATTR_COLD;
+
+	void common_init();
```
If this is only used at start time, it should be marked cold, too.

cuavas reviewed (changes requested):
Please use anonymous namespaces in self-contained driver source files to help keep linking time down, avoid symbol conflicts, etc.

mamehaze commented:
Most of your comments here are about the existing code, that will be
cleaned up later, not in this PR, please stop doing that, it makes
incremental progress impossible.  The code is no worse than it already is
in mame.

Anything specific to what I've done ill fix.

On Fri, 28 Feb 2025, 13:52 Vas Crabb, ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
>
> Please use anonymous namespaces in self-contained driver source files to
> help keep linking time down, avoid symbol conflicts, etc.
> ------------------------------
>
> In src/mame/tatsumi/cyclwarr.cpp
> <https://github.com/mamedev/mame/pull/13418#discussion_r1975411717>:
>
> > +    - (fixed) Cycle Warriors: transparent road layer on sidelines, wrong mask_data?
> +    - (fixed) Missing BG layer (Round Up 5) - banked VRAM data from somewhere!?
> +    - (fixed?) Cycle Warriors: test mode text does not appear as it needs a -256 Y
> +      scroll offset from somewhere.
> +    - (fixed) Cycle Warriors: sometimes it draws garbage on character select or even hangs
> +      depending on where player coins up, most likely caused by miscommunication with sub CPU?
> +    - (fixed) Cycle Warriors: ranking screen is completely wrong;
> +    - (fixed) Cycle Warriors: ugly orange color on character select and briefing screens, layer disable?
>
> There’s no point keeping the stuff that’s done in the TODO list…
> ------------------------------
>
> In src/mame/tatsumi/cyclwarr.cpp
> <https://github.com/mamedev/mame/pull/13418#discussion_r1975412630>:
>
> > +	tileinfo.set(0,
> +						 tileno,
> +						 color,
> +						opaque ? TILE_FORCE_LAYER0 : 0);
>
> Indentation is all over the place here.
> ------------------------------
>
> In src/mame/tatsumi/cyclwarr.cpp
> <https://github.com/mamedev/mame/pull/13418#discussion_r1975413304>:
>
> > +	tileinfo.set(0,
> +						 tileno,
> +						 color | m_road_color_bank,
> +						 opaque ? TILE_FORCE_LAYER0 : 0);
>
> Same here.
> ------------------------------
>
> In src/mame/tatsumi/roundup5.cpp
> <https://github.com/mamedev/mame/pull/13418#discussion_r1975417275>:
>
> > +	DECLARE_VIDEO_START(roundup5);
> +	uint32_t screen_update_roundup5(screen_device &screen, bitmap_rgb32 &bitmap, const rectangle &cliprect);
>
> You should just override video_start() for this derived class rather than
> using the legacy VIDEO_START stuff.
> ------------------------------
>
> In src/mame/tatsumi/roundup5.cpp
> <https://github.com/mamedev/mame/pull/13418#discussion_r1975424146>:
>
> > +void roundup5_state::roundup5_control_w(offs_t offset, uint16_t data, uint16_t mem_mask)
> +{
> +	COMBINE_DATA(&m_control_word);
> +
> +	if (m_control_word & 0x10)
> +		m_subcpu->set_input_line(INPUT_LINE_HALT, ASSERT_LINE);
> +	else
> +		m_subcpu->set_input_line(INPUT_LINE_HALT, CLEAR_LINE);
> +
> +	if (m_control_word & 0x4)
> +		m_audiocpu->set_input_line(INPUT_LINE_HALT, ASSERT_LINE);
> +	else
> +		m_audiocpu->set_input_line(INPUT_LINE_HALT, CLEAR_LINE);
> +
> +//  if (offset == 1 && (tatsumi_control_w & 0xfeff) != (last_bank & 0xfeff))
> +//      logerror("%s:  Changed bank to %04x (%d)\n", m_maincpu->pc(), tatsumi_control_w,offset);
> +
> +//todo - watchdog
> +
> +	/* Bank:
> +
> +	    0x0017  :   OBJ banks
> +	    0x0018  :   68000 RAM       mask 0x0380 used to save bits when writing
> +	    0x0c10  :   68000 ROM
> +
> +	    0x0040  :   Z80 rom (lower half) mapped to 0x10000
> +	    0x0060  :   Z80 rom (upper half) mapped to 0x10000
> +
> +	    0x0080  :   enabled when showing map screen after a play
> +	                (switches video priority between text layer and sprites)
> +
> +	    0x0100  :   watchdog.
> +
> +	    0x0c00  :   vram bank (bits 0x7000 also set when writing vram)
> +
> +	    0x8000  :   set whenever writing to palette ram?
> +
> +	    Changed bank to 0060 (0)
> +	*/
> +
> +	if ((m_control_word & 0x8) == 0 && !(m_last_control & 0x8))
> +		m_subcpu->set_input_line(INPUT_LINE_IRQ4, ASSERT_LINE);
> +//  if (tatsumi_control_w&0x200)
> +//      cpu_set_reset_line(1, CLEAR_LINE);
> +//  else
> +//      cpu_set_reset_line(1, ASSERT_LINE);
> +
> +//  if ((tatsumi_control_w&0x200) && (last_bank&0x200)==0)
> +//      logerror("68k irq\n");
> +//  if ((tatsumi_control_w&0x200)==0 && (last_bank&0x200)==0x200)
> +//      logerror("68k reset\n");
> +
> +	if (m_control_word == 0x3a00)
> +	{
> +//      cpu_set_reset_line(1, CLEAR_LINE);
> +//      logerror("68k on\n");
> +	}
> +
> +	m_last_control = m_control_word;
> +}
>
> This is pretty gross.
>
> m_last_control will always equal m_control_word outside this function, so
> it’s pointless making both of them members. It would be more sensible to
> make the new control word value a local variable, and update
> m_control_word at the end.
>
> Also, it’s going to be going through the scheduler every time the halt
> lines for m_audiocpu and m_subcpu *don’t* change because it isn’t
> checking that they’ve actually changed.
> ------------------------------
>
> In src/mame/tatsumi/tatsumi_rotating_sprites.h
> <https://github.com/mamedev/mame/pull/13418#discussion_r1975432592>:
>
> > +
> +	DECLARE_GFXDECODE_MEMBER(gfxinfo);
>
> Does this need to be public?
> ------------------------------
>
> In src/mame/tatsumi/tatsumi_rotating_sprites.h
> <https://github.com/mamedev/mame/pull/13418#discussion_r1975433888>:
>
> > +	std::unique_ptr<uint8_t[]> m_shadow_pen_array;
> +
> +	bitmap_rgb32 m_temp_bitmap;
> +
> +	int m_sprite_palette_base;
> +
> +	void mycopyrozbitmap_core(bitmap_ind8 &bitmap, const bitmap_rgb32 &srcbitmap,
> +			int dstx, int dsty, int srcwidth, int srcheight, int incxx, int incxy, int incyx, int incyy,
> +			const rectangle &clip, int transparent_color);
> +	void mycopyrozbitmap_core(bitmap_rgb32 &bitmap, const bitmap_rgb32 &srcbitmap,
> +			int dstx, int dsty, int srcwidth, int srcheight, int incxx, int incxy, int incyx, int incyy,
> +			const rectangle &clip, int transparent_color);
> +
> +	template<class BitmapClass> void draw_sprites_main(BitmapClass &bitmap, const rectangle &cliprect, int write_priority_only, int rambank);
> +	template<class BitmapClass> inline void roundupt_drawgfxzoomrotate( BitmapClass &dest_bmp, const rectangle &clip,
> +		gfx_element *gfx, uint32_t code,uint32_t color,int flipx,int flipy,uint32_t ssx,uint32_t ssy,
> +		int scalex, int scaley, int rotate, int write_priority_only );
> +
> +	required_shared_ptr<uint16_t> m_spriteram;
> +	required_region_ptr<uint8_t> m_sprites_l_rom;
> +	required_region_ptr<uint8_t> m_sprites_h_rom;
>
> Please keep the data members together and the member functions together.
> ------------------------------
>
> In src/mame/tatsumi/tatsumi_rotating_sprites.cpp
> <https://github.com/mamedev/mame/pull/13418#discussion_r1975443716>:
>
> > +void tatsumi_rotating_sprites_device::device_add_mconfig(machine_config &config)
> +{
> +	PALETTE(config, m_fakepalette).set_format(palette_device::xRGB_555, 4096); // 4096 arranged as series of CLUTs
> +}
> +
> +void tatsumi_rotating_sprites_bigpal_device::device_add_mconfig(machine_config &config)
> +{
> +	PALETTE(config, m_fakepalette).set_format(palette_device::xRGB_555, 8192); // 8192 arranged as series of CLUTs
> +}
>
> You don’t need to override this in the derived class, you can just use m_rom_clut_size
> * 2 for the number of entries.
> ------------------------------
>
> In src/mame/tatsumi/tatsumi_rotating_sprites.h
> <https://github.com/mamedev/mame/pull/13418#discussion_r1975446363>:
>
> > +	void draw_sprites(bitmap_rgb32& bitmap, const rectangle& cliprect, int write_priority_only, int rambank);
> +	void draw_sprites(bitmap_ind8& bitmap, const rectangle& cliprect, int write_priority_only, int rambank);
> +
> +	void update_cluts();
> +
> +	DECLARE_GFXDECODE_MEMBER(gfxinfo);
> +
> +protected:
> +	tatsumi_rotating_sprites_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock);
> +
> +	virtual void device_add_mconfig(machine_config &config) override ATTR_COLD;
> +
> +	virtual void device_start() override ATTR_COLD;
> +	virtual void device_reset() override ATTR_COLD;
> +
> +	void common_init();
>
> If this is only used at start time, it should be marked cold, too.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mamedev/mame/pull/13418#pullrequestreview-2650748872>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBR6GZNQ7VT47IYSY2ETZLD2SBS3LAVCNFSM6AAAAABX25XDTCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMNJQG42DQOBXGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

mamehaze commented:
the changes that are going to be addressed have been.  anything that might result in a functional change, however minor, has not been altered.

cuavas commented:
Using commit messages as a platform for personal antagonisation is not acceptable behaviour.

mamehaze commented:
Seriously, fuck you

mamehaze commented:
I'd done as you've demanded, so just merge the fucking PR already, Jesus Christ.

Or do I have to revert all the moving about just to keep you happy, and stop you complaining about code I haven't fucking changed.

cuavas commented:
> Seriously fuck you Vas, you antagonise all the time with your unreasonable demands piled on PRs.

From a person constantly complaining about “rudeness”…

You should definitely know better.  And that aside, commit messages are “on the record” and effectively permanent.  They’re to describe what the change affects, and what it does.  They aren’t for completely unrelated complaints or attacks.

mamehaze commented:
The commit messages are accurate, the changes I'm being forced to make are out of scope BS for the PR in question.  If you want them labelled any other way, don't make out of scope demands and trust your developers to get on with things.

cuavas commented:
> I'd done as you've demanded, so just merge the fucking PR already, Jesus Christ.

This is not acceptable directed at anyone.  I’ve put up with insults and attacks when they’re directed at me, but you just keep trying to push it further.

> Or do I have to revert all the moving about just to keep you happy, and stop you complaining about code I haven't fucking changed.

If you’re moving stuff around you should at least be fixing obvious things as you go.

mamehaze commented:
> If you’re moving stuff around you should at least be fixing obvious things as you go.

I'm intentionally changing as little as possible at this stage.  That is my work flow.  Get things where I want them, clean up later.

Some of your demands would even potentially result in behavior changes, which I *absolutely* want to avoid at this stage.

Nothing I'm doing is making MAME worse, you're not judging things on the merits of the PR here, which are overall an improvement, you're making bullshit demands and I'll call them out as such.

happppp commented:
"Squash and merge" can get rid of all the useless commit messages like "shuffle some bits around" or the one calling cuavas's feedback BS on code that was already in MAME.

cuavas commented:
> "Squash and merge" can get rid of all the useless commit messages like "shuffle some bits around" or the one calling cuavas's feedback BS on code that was already in MAME.

It doesn’t get rid of them from the commit log on the PR, and it doesn’t change the fact that it’s unacceptable.  Descriptive commit messages is a pretty low bar.

mamehaze commented:
"Fix out of scope bullshit" is pretty descriptive when I've been asked to fix out of scope bullshit.

Judging PRs on their own merits is also a very low bar to ask, but you're never happy with something that makes MAME better overall, you insist on micromanaging it all, insist on changes that are not ready to be made, insist on changes being made to code that has very intentionally not been changed yet.

If the PR never gets merged, the code you're objecting to is still in MAME and will likely remain in MAME because it's been in MAME for 20 odd years.  The PR isn't about changing that yet, that's the next stage.
